internal/testutil/totp/totp_test.go

```diff
@@ -0,0 +1,44 @@
+package totp
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestGenerateOTPCode(t *testing.T) {
+	tests := []struct {
+		name                         string
+		token                        string
+		when                         time.Time
+		wantCode                     string
+		wantRemainingLifetimeSeconds int64
+	}{
+		{
+			name:                         "Use a token from online example",
+			token:                        "JBSWY3DPEHPK3PXP", // https://github.com/pquerna/otp/blob/3357de7c04813a328d6a1e4a514854213e0f8ce8/totp/totp.go#L180
+			when:                         time.Unix(1715205169, 0),
+			wantCode:                     "780919",
+			wantRemainingLifetimeSeconds: 11,
+		},
+		{
+			name:                         "Use a token that was randomly generated",
+			token:                        "EDAYKXL3TEYZNQ3O4N5KPSUAQQLZYUJG",
+			when:                         time.Unix(1715225917, 0),
+			wantCode:                     "920615",
+			wantRemainingLifetimeSeconds: 23,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			actualCode, actualRemainingLifetimeSeconds := GenerateOTPCode(t, test.token, test.when)
+
+			require.Equal(t, test.wantCode, actualCode)
+			require.Equal(t, test.wantRemainingLifetimeSeconds, actualRemainingLifetimeSeconds)
```

Comment:
Should this assertions be within an acceptable delta? On a slow CI worker running lots of parallel tests, I'd imagine that this could slip into the next second and be off by 1.

Reply:
It should be ok to leave as-is. I don't think the code-under-test makes use of system time at all, it just uses the time that's passed in, and the `remainingLifetimeSeconds` is just how much time is left in that 30-second window, which should not be machine dependent.

Reply:
Ah, ok. Thanks for checking.